### PR TITLE
updated to PureScript 0.12

### PR DESCRIPTION
**Pull request by CarstenKoenig**
## What does this pull request do?

- updated dependencies
- replaced removed `Generic.Rep` cases/fields with a `RowToList` approach

## Where should the reviewer start?

the relevant changes should be in `Data.Argonaut.Decode.Generic.Rep` and `Data.Argonaut.Encode.Generic.Rep`

## How should this be manually tested?

    npm run test

or

    pulp test

should do

## Other Notes:

this is a cleaned up version from this PR: https://github.com/purescript-contrib/purescript-argonaut-generic/pull/6#pullrequestreview-132073440

### CODE REVIEW

File: src/Data/Argonaut/Decode/Generic/Rep.purs [hunk 1]
```diff
@@ -63,28 +69,76 @@ instance decodeRepArgsProduct :: (DecodeRepArgs a, DecodeRepArgs b) => DecodeRep
...
+-- |
+-- | a `RowList` on the type level is very similar to a *cons-list* on the value level
+-- | so the two instances handle all possible `RowList`s
+-- |
+-- | the idea is to use `Builder` to convert a `RowList` into a record at the type-level
+-- | and have `decodeRepRowList` as witness on the value level that will try to decode
+-- | JSON in to the resulting record value
+-- | 
+-- | `from` and `to` indicates the the step the `Builder` will take from a `Record from`
+-- | to a `Record to`  so for a given `RowList` `from` will be the empty record
```
Comment (fsoikin) [inline src/Data/Argonaut/Decode/Generic/Rep.purs]: If `from` is always an empty record, then why is it variable here?

Comment (CarstenKoenig) [reply src/Data/Argonaut/Decode/Generic/Rep.purs]: maybe I put this wrong - I think it should help if you take the class as some kind of signature and the two instances as the implementation

the *nil* cases fixes both to `()` (empty row) and *cons* case will prepend a symbol by using one in the middle (the `from'` in there)

so when everything is resolved you will have a `Builder () rec` (not exactly the right syntax but I hope you get it) - a record-builder that can build the target record from an empty one (which is what we want here: https://github.com/purescript-contrib/purescript-argonaut-generic/pull/8/files#diff-fdc5c38e890eb2d992cb507448e4ce91R80 - here we take the record-builder-steps `decodeRepRowList` (you could say the instance-resolution) found us and use those to build up the record form `{}`)

it's a bit like those graph algorithms where you are looking for a shortest path or something from A to B and you get it by picking points C in between and search for A -> C -> B recursively - here the way is very linear of course but you have to walk along step by step anyway

I think I would probably have to try harder to explain this - but I think there is a purescript hangout video about this stuff too (AFAIK coming from Justin Woo - actually I copy&pasted most of this from one of his examples/libs - don't remember which one right now - I always struggle to remember the details and right now it's hard to find the `RowList` defintions etc. on pursuit)

Comment (fsoikin) [reply src/Data/Argonaut/Decode/Generic/Rep.purs]: I think there is a misunderstanding.

I do understand how this works. I've already spent time to understand it, and also I happen to have written similar things multiple times before.

The comment is not for me, but for future audience. Maintainability, you know.

Comment (CarstenKoenig) [reply src/Data/Argonaut/Decode/Generic/Rep.purs]: oh shit sorry - maybe you can help me out with the wording here then

Comment (CarstenKoenig) [reply src/Data/Argonaut/Decode/Generic/Rep.purs]: I'm really only trying to help - of course you know this ...

anyway can you have a look at it now? I hope the confusing pieces are gone and it's more obvious why to use this(?)

Comment (fsoikin) [reply src/Data/Argonaut/Decode/Generic/Rep.purs]: Don't worry too much about it, I only suggested. 
I think this comment is great.

**Review by fsoikin — APPROVED** [review]

**Comment by thomashoneyman** [comment]
@fsoikin Since you've got an approved review in, I'll merge this tomorrow unless you have any objections. Thanks @CarstenKoenig for your work!

**Comment by CarstenKoenig** [comment]
thank you!

**Comment by fsoikin** [comment]
No objections, please merge. Thank you! :-)

**Comment by thomashoneyman** [comment]
👍

**Pull request merged**